test/unit/compiler/passes/helpers.js

```diff
@@ -46,6 +46,11 @@ module.exports = function(chai, utils) {
 
     let ast = parser.parse(grammar);
 
+    if (!options.allowedStartRules) {
+      options.allowedStartRules = ast.rules.length > 0
+        ? [ast.rules[0].name]
+        : [];
+    }
```

Comment:
> `ast.rules.length` is always 1 or more

`parse` contract not guarantee it. Grammar without any rules still valid and can be occured in tests. The same reason for which all this `if` block in general exists -- tests are the special code demanding the special relation.

Reply:
What do you mean the parse contract doesn't guarantee it? The [parser](/pegjs/pegjs/blob/master/lib/parser.js) currently generated from the [grammar](/pegjs/pegjs/blob/master/src/parser.pegjs) will throw an error when a grammar without any rules is passed to it, so it is guaranteed that there will always be one rule from the PEG.js parser.

As I could not find a test case that ensured empty grammars are not accepted by the parser, I [added one](/pegjs/pegjs/commit/abadaf4d5d9db5fc0c476de51037bd7fadda678d) on my [development-overhaul branch](/pegjs/pegjs/tree/development-overhaul).

As for the compiler option `allowedStartRules`, since there will always be at least 1 rule, you can always expect it to be set as the compiler option default for `allowedStartRules`, so there's no need to set this, even in the test's, unless you are intentionally passing a rule name that doesn't exist to ensure the compiler throws an error.

Reply:
> The parser currently generated from the grammar will throw an error when a grammar without any rules is passed to it

This is implementation detail. There are no reasonable reasons to do it. And if it changes in the future, tests which do not depend on it shall not fall. To test in the assumption as the code works means that you have no tests.

> so there's no need to set this, even in the test's,

As it is [possible to see few lines above][1], the `options` can be not supplied to the test and then it is meant empty. Thus, malformed data which to it shall not be supplied are obviously supplied to pass. Whether means it that the test in principle is totally wrong because it tries to expect correct behavior from a code, supply it certainly malformed data?

From this point of view this code only adds reasonable defaults to a test code, without forcing the developer of tests to supply their every time. Such simplification of operation is also required from the utility method, you do not agree?

[1]: https://github.com/pegjs/pegjs/pull/399/files/375cc1b854cefeeda122cd96d0ac5e16a09b779f#diff-55e8117a5d87da917236f6c83423e206R9

Reply:
Since I do plan to add module support (#523) and annotations, it is likely in the future that there will be grammar files without rules, and you are correct that having this in the helper could actually be a plus for any developers making tests, so I accept this change, _but_ can you extract this into a separate PR as this is a separate issue from this PR.

While you're at it, could you also in this new PR remove the related `if ( options.allowedStartRules ) ` in [lib/compiler/passes/report-undefined-rules.js#L22](https://github.com/pegjs/pegjs/blob/master/lib/compiler/passes/report-undefined-rules.js#L22)

If you don't want to make the new PR, just tell me and remove this code snippet from this PR, then I'll do the new PR at a later date.